File: src/Pybind11Wraps/NodeList/SolidNodeList.py
from PYB11Generator import *
from FluidNodeList import FluidNodeList

#-------------------------------------------------------------------------------
# SolidNodeList template
#-------------------------------------------------------------------------------
@PYB11template("Dimension")
class SolidNodeList(FluidNodeList):
    "Spheral SolidNodeList base class in %(Dimension)s, i.e.,  the NodeList for solid dynamics."

    typedefs = """
    typedef typename %(Dimension)s::Scalar Scalar;
    typedef typename %(Dimension)s::Vector Vector;
    typedef typename %(Dimension)s::Tensor Tensor;
    typedef typename %(Dimension)s::SymTensor SymTensor;
    typedef Field<%(Dimension)s, int> IntField;
    typedef Field<%(Dimension)s, Scalar> ScalarField;
    typedef Field<%(Dimension)s, Vector> VectorField;
    typedef Field<%(Dimension)s, Tensor> TensorField;
    typedef Field<%(Dimension)s, SymTensor> SymTensorField;
"""

    def pyinit(self,
               name = "std::string",
               eos = "EquationOfState<%(Dimension)s>&",
               strength = "StrengthModel<%(Dimension)s>&",
               numInternal = ("int", "0"),
               numGhost = ("int", "0"),
               hmin = ("double", "1e-20"),
               hmax = ("double", "1e20"),
               hminratio = ("double", "0.1"),
               nPerh = ("double", "2.01"),
               maxNumNeighbors = ("int", "500"),
               rhoMin = ("double", "1e-10"),
               rhoMax = ("double", "1e100")):
        "Constructor for a SolidNodeList class."
        return

    @PYB11virtual
    @PYB11const
    def soundSpeed(self, result="ScalarField&"):
        "Compute the current sound speed, storing the result in the argument ScalarField."
        return "void"

    @PYB11virtual
    @PYB11const
    def bulkModulus(self, result="ScalarField&"):
        "Compute the current bulk modulus, storing the result in the argument ScalarField."
        return "void"

    @PYB11virtual
    @PYB11const
    def yieldStrength(self, result="ScalarField&"):
        "Compute the current yield strength, storing the result in the argument ScalarField."
        return "void"

    @PYB11const
    def deviatoricStress(self):
        "The mass density field"
        return "const SymTensorField&"

    @PYB11const
    def plasticStrain(self):
        "The plastic strain field"
        return "const ScalarField&"

    @PYB11const
    def plasticStrainRate(self):
        "The plastic strain rate field"
        return "const ScalarField&"

    @PYB11const
    def damage(self):
        "The damage field"
        return "const SymTensorField&"

    @PYB11const
    def effectiveDamage(self):
        "The effective damage field"
        return "const SymTensorField&"

    @PYB11const
    def damageGradient(self):
        "The damage gradient field"
        return "const VectorField&"

    @PYB11const
    def fragmentIDs(self):
        "The fragment IDs field"
        return "const IntField&"

    @PYB11const
    def particleTypes(self):
        "The particle type field"
        return "const IntField&"

    @PYB11const
    def strengthModel(self):
        "Return the strength model object this SolidNodeList is associated with."
        return "const StrengthModel<%(Dimension)s>&"

    @PYB11virtual
    @PYB11const
    def label(self):
        "Label for restart files"
        return "std::string"

    @PYB11virtual
    @PYB11const
    def dumpState(self, file="FileIO&", pathName="const std::string&"):
        "Serialize under the given path in a FileIO object"
        return "void"

    @PYB11virtual
    def restoreState(self, file="const FileIO&", pathName="const std::string&"):
        "Restore state from the given path in a FileIO object"
        return "void"

    # Comparison
    def __eq__(self):
        "Equivalence test with another SolidNodeList"

    def __ne__(self):
        "Inequivalence test with another SolidNodeList"

File: src/NodeList/SolidNodeList.cc
//---------------------------------Spheral++----------------------------------//
// SolidNodeList -- A form of the SPH NodeList appropriate for use with 
// solid materials.
//
// Created by JMO, Tue Sep 7 22:44:37 2004
//----------------------------------------------------------------------------//
#include "FileIO/FileIO.hh"
#include "Field/Field.hh"
#include "Field/FieldList.hh"
#include "Material/EquationOfState.hh"
#include "Kernel/TableKernel.hh"
#include "Hydro/HydroFieldNames.hh"
#include "Strength/SolidFieldNames.hh"
#include "DataBase/IncrementState.hh"
#include "SolidMaterial/StrengthModel.hh"
#include "Utilities/safeInv.hh"
#include "Utilities/DBC.hh"

#include "SolidNodeList.hh"

using std::vector;
using std::list;
using std::string;
using std::cerr;
using std::endl;

namespace Spheral {

//------------------------------------------------------------------------------
// Construct with the given EOS object, along with optional numInternal nodes,
// numGhost nodes, and name.
//------------------------------------------------------------------------------
template<typename Dimension>
SolidNodeList<Dimension>::
SolidNodeList(string name,
              EquationOfState<Dimension>& eos,
              StrengthModel<Dimension>& strength,
              const size_t numInternal,
              const size_t numGhost,
              const Scalar hmin,
              const Scalar hmax,
              const Scalar hminratio,
              const Scalar nPerh,
              const size_t maxNumNeighbors,
              const Scalar rhoMin,
              const Scalar rhoMax):
  FluidNodeList<Dimension>(name, 
                           eos,
                           numInternal, 
                           numGhost,
                           hmin,
                           hmax,
                           hminratio,
                           nPerh,
                           maxNumNeighbors,
                           rhoMin,
                           rhoMax),
  mDeviatoricStress(SolidFieldNames::deviatoricStress, *this),
  mPlasticStrain(SolidFieldNames::plasticStrain, *this),
  mPlasticStrainRate(SolidFieldNames::plasticStrainRate, *this),
  mDamage(SolidFieldNames::tensorDamage, *this),
  mFragmentIDs(SolidFieldNames::fragmentIDs, *this),
  mParticleTypes(SolidFieldNames::particleTypes, *this),
  mStrength(strength) {
}

//------------------------------------------------------------------------------
// Calculate and return the sound speed field.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
soundSpeed(Field<Dimension, typename Dimension::Scalar>& field) const {

  // Get the straight EOS look up from the base class.
  FluidNodeList<Dimension>::soundSpeed(field);

  // Augment the sound speed with the strength model.
  const auto& rho = this->massDensity();
  const auto& u = this->specificThermalEnergy();
  const auto& D = this->damage();
  Field<Dimension, Scalar> P(HydroFieldNames::pressure, *this);
  this->pressure(P);
  mStrength.soundSpeed(field, rho, u, P, field, D);
}

//------------------------------------------------------------------------------
// Calculate and return the bulk modulus.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
bulkModulus(Field<Dimension, typename Dimension::Scalar>& field) const {
  const auto& rho = this->massDensity();
  const auto& u = this->specificThermalEnergy();
  this->equationOfState().setBulkModulus(field, rho, u);
}

//------------------------------------------------------------------------------
// Calculate and return the shear modulus.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
shearModulus(Field<Dimension, typename Dimension::Scalar>& field) const {
  const auto& rho = this->massDensity();
  const auto& u = this->specificThermalEnergy();
  const auto& D = this->damage();
  Field<Dimension, Scalar> P(HydroFieldNames::pressure, *this);
  this->pressure(P);
  mStrength.shearModulus(field, rho, u, P, D);
}

//------------------------------------------------------------------------------
// Calculate and return the yield strength.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
yieldStrength(Field<Dimension, typename Dimension::Scalar>& field) const {
  const auto& rho = this->massDensity();
  const auto& u = this->specificThermalEnergy();
  const auto& D = this->damage();
  Field<Dimension, Scalar> P(HydroFieldNames::pressure, *this);
  this->pressure(P);
  mStrength.yieldStrength(field, rho, u, P, mPlasticStrain, mPlasticStrainRate, D);
}

//------------------------------------------------------------------------------
// Calculate and return Youngs modulus.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
YoungsModulus(Field<Dimension, typename Dimension::Scalar>& field,
              const Field<Dimension, typename Dimension::Scalar>& K,          // bulk modulus
              const Field<Dimension, typename Dimension::Scalar>& mu) const { // shear modulus
  REQUIRE(K.nodeList().name() == this->name());
  REQUIRE(mu.nodeList().name() == this->name());
  const auto n = this->numInternalNodes();
#pragma omp parallel for
  for (auto i = 0u; i < n; ++i) {
    field(i) = 9.0*K(i)*mu(i)*safeInv(3.0*K(i) + mu(i));
  }
}

//------------------------------------------------------------------------------
// Calculate and return the longitudinal sound speed
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
longitudinalSoundSpeed(Field<Dimension, typename Dimension::Scalar>& field,
                       const Field<Dimension, typename Dimension::Scalar>& rho,        // mass density
                       const Field<Dimension, typename Dimension::Scalar>& K,          // bulk modulus
                       const Field<Dimension, typename Dimension::Scalar>& mu) const { // shear modulus
  REQUIRE(rho.nodeList().name() == this->name());
  REQUIRE(K.nodeList().name() == this->name());
  REQUIRE(mu.nodeList().name() == this->name());
  const auto n = this->numInternalNodes();
#pragma omp parallel for
  for (auto i = 0u; i < n; ++i) {
    field(i) = std::sqrt(std::abs(K(i) + 4.0/3.0*mu(i))*safeInv(rho(i)));
    // const auto ack = 3.0*K(i) + mu(i);
    // const auto nu = std::min(0.5, std::max(0.0, 0.5*(3.0*K(i) - 2.0*mu(i))*safeInv(ack)));
    // CHECK(nu >= 0.0 and nu <= 0.5);
    // const auto barf = (1.0 + nu)*(1.0 - 2.0*nu) + 1.0e-10;
    // CHECK(distinctlyGreaterThan(barf, 0.0));
    // field(i) = std::sqrt(std::abs(E(i)*(1.0 - nu)*safeInv(rho(i)*barf)));
    ENSURE(field(i) >= 0.0);
  }
}

//------------------------------------------------------------------------------
// Dump the current state of the NodeList to the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
dumpState(FileIO& file, const string& pathName) const {

  // Dump the ancestor class.
  FluidNodeList<Dimension>::dumpState(file, pathName);

  file.write(mDeviatoricStress, pathName + "/" + mDeviatoricStress.name());
  file.write(mPlasticStrain, pathName + "/" + mPlasticStrain.name());
  file.write(mPlasticStrainRate, pathName + "/" + mPlasticStrainRate.name());
  file.write(mDamage, pathName + "/" + mDamage.name());
  file.write(mFragmentIDs, pathName + "/" + mFragmentIDs.name());
  file.write(mParticleTypes, pathName + "/" + mParticleTypes.name());
}

//------------------------------------------------------------------------------
// Restore the state of the NodeList from the given file.
//------------------------------------------------------------------------------
template<typename Dimension>
void
SolidNodeList<Dimension>::
restoreState(const FileIO& file, const string& pathName) {

  // Restore the ancestor class.
  FluidNodeList<Dimension>::restoreState(file, pathName);

  file.read(mDeviatoricStress, pathName + "/" + mDeviatoricStress.name());
  file.read(mPlasticStrain, pathName + "/" + mPlasticStrain.name());
  file.read(mPlasticStrainRate, pathName + "/" + mPlasticStrainRate.name());
  file.read(mDamage, pathName + "/" + mDamage.name());
  file.read(mFragmentIDs, pathName + "/" + mFragmentIDs.name());
  file.read(mParticleTypes, pathName + "/" + mParticleTypes.name());
}

}
